src/redux/selectors/orgUnitDistReport/getTransformedTableData.js

```diff
@@ -0,0 +1,149 @@
+import i18n from '@dhis2/d2-i18n'
+import isEmpty from 'lodash.isempty'
+import createDataTransformCache from '../../../utils/dataTransformCache'
+
+const totalColumnDisplayName = i18n.t('Total')
+export const cache = createDataTransformCache()
+
+/**
+ * @param {Object} state - App state
+ * @returns {Object} - Transformed table data that can be consumed by TabularReport
+ */
+export default function getTransformedTableData(state) {
+    const data = state.reportData.content
+
+    if (isEmpty(data)) {
+        return data
+    }
+
+    if (cache.hasValidCacheFor(data)) {
+        return cache.getCachedResult()
+    }
+
+    const grid = createGrid(data)
+    const tableData = {
+        title: getTitle(state),
+        headers: getHeaders(data.headers, grid.columns),
+        rows: getRows(state, grid),
+    }
+
+    cache.setCachedResult(data, tableData)
+    return tableData
+}
+
+/**
+ * @param {Object} data - Response data from the server
+ * @returns {Object} - Containing rows and columns array, columns sorted ASC
+ */
+function createGrid(data) {
+    // As new Set for constant time look-up
+    const orgUnitLookup = new Set(data.orgUnitIds)
+    const items = data.metaData.items
+    const { rows, columns } = Object.keys(items).reduce(
```

Comment:
I've seen this style of transforming an object quite a few times now. We could create a utility function for this called `reduceObject`. Would render the code more readable, sth like:

```
export const reduceObject = (callback, obj, initialData = {}) => Object.keys(obj).reduce(
  (acc, cur) => callback(acc, cur, obj),
  initialData,
)
```

Then you can do:
```
const { rows, columns } = reduceObject(
    (acc, id, items) => {
        const targetArray = orgUnitLookup.has(id) ? acc.rows : acc.columns
        targetArray.push({ id: id, name: items[id].name })
        return acc
    },
    items,
    { rows: [], columns: [] },
);
```

But not really necessary, just an idea to abstract away the Object.keys call. Take this idea or leave it :)

Reply:
Good suggestion 👍 I will keep this in mind for the future. What I was planning to do was the following:
- Work on the data transformations for all sections
- Review all side by side and extract shared logic
Your suggestion would definitely be a good candidate for something to extract....